Userland/Libraries/LibSQL/SQLClient.cpp

```diff
@@ -5,10 +5,132 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/DeprecatedString.h>
 #include <LibSQL/SQLClient.h>
 
+#if !defined(AK_OS_SERENITY)
+#    include <LibCore/Directory.h>
+#    include <LibCore/File.h>
+#    include <LibCore/SocketAddress.h>
+#    include <LibCore/StandardPaths.h>
+#    include <LibCore/Stream.h>
+#    include <LibCore/System.h>
+#endif
+
 namespace SQL {
 
+#if !defined(AK_OS_SERENITY)
+
+// This is heavily based on how SystemServer's Service creates its socket.
+static ErrorOr<int> create_database_socket(DeprecatedString const& socket_path)
+{
+    if (Core::File::exists(socket_path))
+        TRY(Core::System::unlink(socket_path));
+
+#    ifdef SOCK_NONBLOCK
+    auto socket_fd = TRY(Core::System::socket(AF_LOCAL, SOCK_STREAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0));
+#    else
+    auto socket_fd = TRY(Core::System::socket(AF_LOCAL, SOCK_STREAM, 0));
+
+    int option = 1;
+    TRY(Core::System::ioctl(socket_fd, FIONBIO, &option));
+    TRY(Core::System::fcntl(socket_fd, F_SETFD, FD_CLOEXEC));
+#    endif
+
+    TRY(Core::System::fchmod(socket_fd, 0600));
+
+    auto socket_address = Core::SocketAddress::local(socket_path);
+    auto socket_address_un = socket_address.to_sockaddr_un().release_value();
+
+    TRY(Core::System::bind(socket_fd, reinterpret_cast<sockaddr*>(&socket_address_un), sizeof(socket_address_un)));
+    TRY(Core::System::listen(socket_fd, 16));
+
+    return socket_fd;
+}
+
+static ErrorOr<void> launch_server(DeprecatedString const& socket_path, DeprecatedString const& pid_path, char const* server_path)
+{
+    auto server_fd = TRY(create_database_socket(socket_path));
+    auto server_pid = TRY(Core::System::fork());
+
+    if (server_pid == 0) {
+        TRY(Core::System::setsid());
+        TRY(Core::System::signal(SIGCHLD, SIG_IGN));
+        server_pid = TRY(Core::System::fork());
+
+        if (server_pid != 0) {
+            auto server_pid_file = TRY(Core::Stream::File::open(pid_path, Core::Stream::OpenMode::Write));
+            TRY(server_pid_file->write(DeprecatedString::number(server_pid).bytes()));
+
+            exit(0);
+        }
+
+        server_fd = TRY(Core::System::dup(server_fd));
+
+        auto takeover_string = DeprecatedString::formatted("{}:{}", socket_path, server_fd);
+        TRY(Core::System::setenv("SOCKET_TAKEOVER"sv, takeover_string, true));
+
+        char const* argv[] = {
+            "SQLServer",
+            "--pid-file",
+            pid_path.characters(),
+            nullptr,
+        };
+
+        if (execvp(server_path, const_cast<char**>(argv)) < 0)
```

Comment:
I think we have a Core::System::exec wrapper

Reply:
Ah nice, and that looks like it handles platform specifics more nicely, I'll use that. We should be using this in Ladybird too :thinking: